src/deleterows.c

```diff
@@ -0,0 +1,231 @@
+#include "data.table.h"
+
+static void computePrefixSum(const int *keep, int *dest, R_xlen_t n, int nthreads);
+static void compactVectorRaw(SEXP col, const int *dest, const int *keep, R_xlen_t new_nrow, R_xlen_t old_nrow);
+
+// Helper function to make columns resizable for delete by reference
+SEXP allocrow(SEXP dt, R_xlen_t n) {
+  if (!INHERITS(dt, char_datatable))
+    error(_("input to allocrow is not a data.table"));
+
+  if (n < 0)
+    error(_("n must be non-negative in allocrow"));
+
+  if (!xlength(dt)) return dt; // zero-column data.table
+
+  const bool verbose = GetVerbose();
+  int n_modified = 0;
+
+  for (R_xlen_t i = 0; i < length(dt); i++) {
+    SEXP col = VECTOR_ELT(dt, i);
+    if (!isVector(col))
+      error("Cannot make non-vector column %lld resizable", (long long)(i + 1));
```

Comment:
Be sure to mark strings for translation as appropriate

```suggestion
      error(_("Cannot make non-vector column %lld resizable"), (long long)(i + 1));
```